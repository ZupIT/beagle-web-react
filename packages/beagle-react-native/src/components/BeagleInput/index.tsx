/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import React, { FC, useState } from 'react'
import { BeagleTextInputInterface, InputHandler } from 'common/models'
import {
  TextInputProps,
  StyleSheet,
  TextInput,
  KeyboardType,
} from 'react-native'

const keyboardTypes: Record<string, KeyboardType> = {
  NUMBER: 'numeric',
  EMAIL: 'email-address',
}

const BeagleTextInput: FC<BeagleTextInputInterface> = ({
  value: initialValue,
  placeholder,
  disabled,
  readOnly,
  type = 'TEXT',
  onChange,
  onFocus,
  onBlur,
  style,
}) => {

  const [value, setValue] = useState(initialValue)
  const isEditable = !disabled && !readOnly

  const handleEvent = (handler?: InputHandler) => (text: string) => {
    setValue(text)
    return handler && handler({ value: text })
  }

  const handleOnFocus = () => onFocus && onFocus({ value })

  const handleOnBlur = () => onBlur && onBlur({ value })

  const inputProps: TextInputProps = {
    value: value,
    placeholder: placeholder,
    onChangeText: handleEvent(onChange),
    onFocus: handleOnFocus,
    onBlur: handleOnBlur,
    editable: isEditable,
    secureTextEntry: type === 'PASSWORD',
    keyboardType: keyboardTypes[type] || 'default',
  }

  const styleSheet = StyleSheet.create({
    fromBffStyles: {
      ...style,
    },
    defaultStyles: {
      flex: style && style.flex ? Number(style.flex) : 1,
      borderWidth: 1,
      borderColor: '#000000',
      borderStyle: 'solid',
      margin: 5,
      maxHeight: 50,
    },
  })

  return (
    <TextInput
      {...inputProps}
      style={{
        ...styleSheet.defaultStyles,
        ...styleSheet.fromBffStyles,
      }}>
    </TextInput>
  )
}

export default BeagleTextInput
