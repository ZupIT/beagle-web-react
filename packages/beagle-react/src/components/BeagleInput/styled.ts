/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import styled from 'styled-components'

interface StyledTextInputInterface {
  error?: string,
  showError?: boolean,
}

export const errorColor = '#FF0000'

export const StyledInput = styled.input<StyledTextInputInterface>`
  border: ${({ error, showError }) => error && showError ? errorColor : ''};
  outline: ${({ error, showError }) => error && showError ? `auto ${errorColor}` : ''};
`
