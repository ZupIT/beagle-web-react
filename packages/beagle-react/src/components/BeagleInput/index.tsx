/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { FC } from 'react'
import { BeagleTextInputInterface, InputHandler } from 'common/models'
import { InputEvent } from '../types'
import withTheme from '../utils/withTheme'
import BeagleText from '../BeagleText'
import { StyledInput, errorColor } from './styled'

const BeagleInput: FC<BeagleTextInputInterface> = ({
  value,
  placeholder,
  disabled,
  readOnly,
  type = 'TEXT',
  onChange,
  onFocus,
  onBlur,
  style,
  className,
  error,
  showError,
}) => {

  const handleEvent = (handler?: InputHandler) => (event: InputEvent) => {
    if (!handler) return
    handler({ value: event.target.value })
  }

  const showErrorMessage = (() => {
    if (error && showError)
      return <BeagleText
        text={error || ''}
        textColor={errorColor}
        style={{ fontSize: '0.8rem' }}>
      </BeagleText>
  })

  return (
    <>
      <StyledInput
        value={value}
        placeholder={placeholder}
        disabled={disabled}
        readOnly={readOnly}
        type={type}
        onChange={handleEvent(onChange)}
        onBlur={handleEvent(onBlur)}
        onFocus={handleEvent(onFocus)}
        style={{ ...style }}
        className={className}
        error={error}
        showError={showError}
      />
      {showErrorMessage()}
    </>
  )
}

export default withTheme(BeagleInput)
