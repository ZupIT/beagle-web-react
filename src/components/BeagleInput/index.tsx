/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import React, { FC } from 'react'
import { InputEvent, InputHandler, BeagleDefaultComponent, InputType } from '../types'
import withTheme from '../utils/withTheme'
import { StyledInput } from './styled'

export interface TextInputInterface extends BeagleDefaultComponent {
  value: string,
  placeholder?: string,
  disabled?: boolean,
  readOnly?: boolean,
  type?: InputType,
  hidden?: boolean,
  onChange?: InputHandler,
  onFocus?: InputHandler,
  onBlur?: InputHandler,
}

const BeagleInput: FC<TextInputInterface> = ({
  value,
  placeholder,
  disabled,
  readOnly,
  type = 'TEXT',
  hidden,
  onChange,
  onFocus,
  onBlur,
  style,
  className,
}) => {
  const handleEvent = (handler?: InputHandler) => (event: InputEvent) => {
    if (!handler) return
    handler({ value: event.target.value })
  }

  return (
    <StyledInput
      value={value}
      placeholder={placeholder}
      disabled={disabled}
      readOnly={readOnly}
      type={type}
      hidden={hidden}
      onChange={handleEvent(onChange)}
      onBlur={handleEvent(onBlur)}
      onFocus={handleEvent(onFocus)}
      style={style}
      className={className}
    />
  )
}

export default withTheme(BeagleInput)
