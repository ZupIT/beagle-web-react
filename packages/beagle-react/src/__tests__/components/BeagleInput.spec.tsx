/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

// Link.react.test.js
import React from 'react'
import Adapter from 'enzyme-adapter-react-16'
import { configure, mount } from 'enzyme'
import BeagleInput, { TextInputInterface } from '../../components/BeagleInput'

let wrapper: any
const beagleInputPropsMock: TextInputInterface = {
  value: 'Testing',
  placeholder: 'Testing',
  disabled: false,
  readOnly: false,
  type: 'TEXT',
  hidden: false,
  onChange: jest.fn(),
  onFocus: jest.fn(),
  onBlur: jest.fn(),
}

configure({ adapter: new Adapter() })
beforeAll(() => {
  wrapper = mount(<BeagleInput {...beagleInputPropsMock} />)
})

test('Beagle snapshot Input', () => {
  expect(wrapper).toMatchSnapshot()
})

test('Should call on change handler', () => {
  const props: any = wrapper.props()
  wrapper.simulate('change', { value: 'changed' })
  expect(props.onChange).toHaveBeenCalledTimes(1)
})

test('Should call on focus handler', () => {
  const props: any = wrapper.props()
  wrapper.simulate('focus')
  expect(props.onFocus).toHaveBeenCalledTimes(1)
})

test('Should call on blur handler', () => {
  const props: any = wrapper.props()
  wrapper.simulate('blur')
  expect(props.onBlur).toHaveBeenCalledTimes(1)
})

