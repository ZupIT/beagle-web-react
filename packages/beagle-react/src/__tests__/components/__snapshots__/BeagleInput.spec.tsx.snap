// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Beagle snapshot Input 1`] = `
<ComponentWithTheme
  disabled={false}
  onBlur={[MockFunction]}
  onChange={[MockFunction]}
  onFocus={[MockFunction]}
  placeholder="Testing"
  readOnly={false}
  type="TEXT"
  value="Testing"
>
  <BeagleInput
    className=""
    disabled={false}
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    onFocus={[MockFunction]}
    placeholder="Testing"
    readOnly={false}
    type="TEXT"
    value="Testing"
  >
    <styled.input
      aria-disabled={false}
      aria-placeholder="Testing"
      aria-readonly={false}
      className=""
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Testing"
      readOnly={false}
      style={Object {}}
      type="TEXT"
      value="Testing"
    >
      <input
        aria-disabled={false}
        aria-placeholder="Testing"
        aria-readonly={false}
        className="sc-gsTCUz cbPgpe"
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Testing"
        readOnly={false}
        style={Object {}}
        type="TEXT"
        value="Testing"
      />
    </styled.input>
  </BeagleInput>
</ComponentWithTheme>
`;
